### PR TITLE
Do not apply Fix All operations to generated code

### LINKED ISSUE #33507: [Regression] Fix alls have started running in generated code
This was fixed during the Dev15 cycle, but seems to have been regressed again. 

1.
```
> git clone http://github.com/dotnet/project-system
> cd project-system
project-system> git checkout 9bac1e669c67e5d0c44486be595d7ec6323461ca
project-system> build.cmd
project-system> ProjectSystem.sln
```
2. Navigate to `ActiveEditorContextTracker.cs`
3. Drop the code fix over `using System.Threading.Tasks` to **Remove unnecessary usings** -> **Solution**

**Expected:**  (15.9)
Two results:
image

**Actual:** (16.0 Preview 4)
100s of results in auto-generated files:
image

### PR DESCRIPTION
**Pull request by sharwell**
Fixes #33507 

### Customer scenario

A customer applies a Fix All in Project or Fix All in Solution operation for **Remove Unnecessary Usings**, and generated source files (e.g. RESX generated files or WinForms designer code) is changed.

### Bugs this fixes

#33507 

### Workarounds, if any

Apply fixes to documents individually.

### Risk

This is generally a measure our how central the affected code is to adjacent
scenarios and thus how likely your fix is to destabilize a broader area of code

### Performance impact

Negligible improvement. Less work is done in a case where would should not be done.

### Is this a regression from a previous update?

Yes, this is a regression from 15.0.

### Root cause analysis

There were previously no tests for this feature behavior. Tests covering numerous edge cases of this feature behavior have been added.

### How was the bug found?

Internal report (dogfood).

### Test documentation updated?

No.

### CODE REVIEW

File: src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpCodeActions.cs [hunk 1]
```diff
@@ -281,6 +282,140 @@ public int Y2
...
+                "Remove Unnecessary Usings",
+                applyFix: true,
+                fixAllScope: scope);
+
+            Assert.Equal(expectedText, VisualStudio.Editor.GetText());
+
+            VisualStudio.SolutionExplorer.OpenFile(new ProjectUtils.Project(ProjectName), "D.cs");
+            Assert.Equal(generatedSource, VisualStudio.Editor.GetText());
+
+            // Verify that a Fix All in Document in the generated file still does nothing
```
Comment (mavasani) [inline src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpCodeActions.cs]: This seems wrong. We should either not be offering a FixAll in document in generated files OR if we do offer it then it must apply the Fix all instances in the document.

File: src/CodeStyle/Core/CodeFixes/FixAllContextHelper.cs [hunk 2]
```diff
@@ -74,12 +74,13 @@ internal static class FixAllContextHelper
...
 
             return await GetDocumentDiagnosticsToFixAsync(
-                allDiagnostics, projectsToFix, fixAllContext.CancellationToken).ConfigureAwait(false);
+                allDiagnostics, projectsToFix, isGeneratedCode, fixAllContext.CancellationToken).ConfigureAwait(false);
         }
 
         private static async Task<ImmutableDictionary<Document, ImmutableArray<Diagnostic>>> GetDocumentDiagnosticsToFixAsync(
             ImmutableArray<Diagnostic> diagnostics,
             ImmutableArray<Project> projects,
+            Func<Document, CancellationToken, bool> isGeneratedCode,
```
Comment (mavasani) [inline src/CodeStyle/Core/CodeFixes/FixAllContextHelper.cs]: We have had multiple discussions in past about whether generated code should be skipped by default in fix all framework or the fixer should write a custom fixall provider to do such filtering and haven’t come to any consensus. I will try to find old issues and discussions around it. This most certainly needs a design discussion. I am also not sure how this bug meets the Dev16.0 bar currently. Can you clarify how this is a regression from 15.0 given FixAll framework never skipped generated files?

File: src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpCodeActions.cs (hunk 1 above)
Comment (sharwell) [reply src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpCodeActions.cs]: This is a regression test showing preservation of 15.0 behavior in 16.0.

**Comment by mavasani** [comment]
I was able to find the commit that regressed from attached issue, and it was news to me that we had switched FixAll to skip generated files by default even before that commit. In that sense, your PR just restores the functionality, so I will remove the block. I would still be keen to hear how others feel about us offering a FixAll in document in generated files that fixes nothing.

**Comment by sharwell** [comment]
> I would still be keen to hear how others feel about us offering a FixAll in document in generated files that fixes nothing.

I wanted to change the behavior so Fix All operations would work in generated code (#26591), but we need to ensure that, at minimum, changes are not made to generated files when the code fix is initiated in non-generated code.

File: src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpCodeActions.cs [hunk 3]
```diff
@@ -140,7 +140,8 @@ void RaiseSecond()
             VisualStudio.Editor.Verify.TextContains("Second?.");
         }
 
-        [WpfFact(Skip = "https://github.com/dotnet/roslyn/issues/30015")]
+        // DO NOT SKIP THIS TEST. IT MUST PASS.
```
Comment (mavasani) [inline src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpCodeActions.cs]: We should try to retest this test multiple times in the PR - it has shown to be very flaky integration test in the past. It might have been fixed with your recent fixes in integration test infrastructure, so we should make sure it passes consistently.

Comment (sharwell) [reply src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpCodeActions.cs]: I updated this test to use `CriticalWpfFact` instead of `WpfFact`. This is an essential test and the only regression prevention we have for the entire feature area; it is a compilation error to try and skip it.

File: src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpCodeActions.cs (hunk 1 above)
Comment (mavasani) [reply src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpCodeActions.cs]: Well, the behavior itself seems broken - offering users a fix that does nothing, and is by design, is neither intuitive nor a good user experience. I am fine with your change as that just reverts to 15.0 behavior, but we may want to discuss just hiding the FixAll in document option for code fixes offered in generated files in 16.1

File: src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpCodeActions.cs (hunk 3 above)
Comment (mavasani) [reply src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpCodeActions.cs]: Tagging @jinujoseph for this `Exceptions are not made on the basis of reliability for this test.` - This integration test was skipped as it used to fail in half the integration test runs. I don't think we should be enabling it again unless we can prove it is not flaky anymore, regardless of how important the test itself is. Additionally, can we please separate out this change into a separate PR, as this seems to be addressing https://github.com/dotnet/roslyn/issues/30015, which has nothing to do with generated code issue being fixed in this PR.

**Comment by sharwell** [comment]
@mavasani I submitted #33517 as a prerequisite to this in its own pull request.

File: src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpCodeActions.cs (hunk 1 above)
Comment (jasonmalinowski) [reply src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpCodeActions.cs]: Can we add a comment that we believe this behavior is wrong? I strongly detest tests that assert known bad behavior without clear comments tothat effect, because it's confusing as hell when you're trying to fix something, "break" a test because you did the right thing, and then ultimately figure it out that way.

Comment (jasonmalinowski) [reply src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpCodeActions.cs]: What would it take to just fix the behavior instead?

Comment (sharwell) [reply src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpCodeActions.cs]: > What would it take to just fix the behavior instead?

That would be for 16.1 at the earliest

**Comment by sharwell** [comment]
> We should really add some additional comments around the assertion that fix alls in a generated document are offered but don't do anything. 

There are four conditions covered by the tests. Two are tests of behavior with respect to intentional design, and two are tests to detect deviations from the current implemented behavior. All four are now documented for clarity.

File: src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpCodeActions.cs (hunk 1 above)
Comment (sharwell) [reply src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpCodeActions.cs]: > Can we add a comment that we believe this behavior is wrong? 

Comments have been added for clarity.

File: src/CodeStyle/Core/CodeFixes/FixAllContextHelper.cs (hunk 2 above)
Comment (sharwell) [reply src/CodeStyle/Core/CodeFixes/FixAllContextHelper.cs]: ➡️ The regression is described in #33507.

File: src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpCodeActions.cs (hunk 3 above)
Comment (sharwell) [reply src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpCodeActions.cs]: ➡️ These changes are no longer part of this PR.

File: src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpCodeActions.cs [hunk 4]
```diff
@@ -281,6 +281,146 @@ public int Y2
...
+                applyFix: true,
+                fixAllScope: scope);
+
+            Assert.Equal(expectedText, VisualStudio.Editor.GetText());
+
+            VisualStudio.SolutionExplorer.OpenFile(new ProjectUtils.Project(ProjectName), "D.cs");
+            Assert.Equal(generatedSource, VisualStudio.Editor.GetText());
+
+            // Verify that a Fix All in Document in the generated file still does nothing.
+            // ⚠ This is a statement of the current behavior, and not a claim regarding correctness of the design.
```
Comment (mavasani) [inline src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpCodeActions.cs]: Is this worthy of filing a tracking design issue or do we want to wait on customer feedback on this known broken behavior before spending any cycles on it?

Comment (jasonmalinowski) [reply src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpCodeActions.cs]: And how hard is this to just fix? :smile:

Comment (sharwell) [reply src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpCodeActions.cs]: > Is this worthy of filing a tracking design issue or do we want to wait on customer feedback on this known broken behavior before spending any cycles on it?

#2753
